From a36a116a0abf96829b3fd074e43d71519411299f Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Sun, 25 Jan 2026 16:11:18 -0600
Subject: [PATCH 05/12] ntdll: Implement NtDelayExecution relative wait using
 QPC.

Some applications are very timing sensitive.
---
 dlls/ntdll/unix/sync.c | 25 ++++++++++++++++++++-----
 1 file changed, 20 insertions(+), 5 deletions(-)

diff --git a/dlls/ntdll/unix/sync.c b/dlls/ntdll/unix/sync.c
index 084bdecddaf..f34b4e67808 100644
--- a/dlls/ntdll/unix/sync.c
+++ b/dlls/ntdll/unix/sync.c
@@ -2474,16 +2474,31 @@ NTSTATUS WINAPI NtDelayExecution( BOOLEAN alertable, const LARGE_INTEGER *timeou
     {
         for (;;) select( 0, NULL, NULL, NULL, NULL );
     }
-    else
+    else if (timeout->QuadPart < 0)
     {
+        timeout_t when = -timeout->QuadPart, diff;
         LARGE_INTEGER now;
-        timeout_t when, diff;
 
-        if ((when = timeout->QuadPart) < 0)
+        NtQueryPerformanceCounter( &now, NULL );
+        when += now.QuadPart;
+
+        NtYieldExecution();
+
+        for (;;)
         {
-            NtQuerySystemTime( &now );
-            when = now.QuadPart - when;
+            struct timeval tv;
+            NtQueryPerformanceCounter( &now, NULL );
+            diff = (when - now.QuadPart + 9) / 10;
+            if (diff <= 0) break;
+            tv.tv_sec = diff / 1000000;
+            tv.tv_usec = diff % 1000000;
+            if (select( 0, NULL, NULL, NULL, &tv ) != -1) break;
         }
+    }
+    else
+    {
+        timeout_t when = timeout->QuadPart, diff;
+        LARGE_INTEGER now;
 
         /* Note that we yield after establishing the desired timeout, but
            we only care about the result of the yield for zero timeouts */

-- 
2.54.0

